pkg/controller/daemon/update.go

```diff
@@ -99,6 +99,7 @@ func (dsc *DaemonSetsController) rollingUpdate(ctx context.Context, ds *apps.Dae
 						allowedReplacementPods = make([]string, 0, len(nodeToDaemonPods))
 					}
 					allowedReplacementPods = append(allowedReplacementPods, oldPod.Name)
+					numUnavailable++
```

Comment:
Incrementing here means enough old unavailable pods can kick us into the `numUnavailable >= maxUnavailable` case and stop us accumulating remaining old unavailable pods into `allowedReplacementPods`.

I agree there was a bug, but I'm not sure this is the best fix ... I think it can dramatically slow down how the controller cleans up old unavailable pods. Before this change, if we had the following scenario:
* maxUnavailable=5
* 4 unavailable new pods
* 100 unavailable old pods
* 1000 available old pods

I _think_ the controller would delete all 100 unavailable old pods and 1 available old pod in a single sync.

With this change, I think it would only delete a single unavailable old pod in each sync loop.

Reply:
I _think_ something like computing `remainingUnavailable := maxUnavailable - numUnavailable - len(allowedReplacementPods)` below may have been a better minimal fix that would have kept the cleanup of all old unavailable pods, while still considering those as filling unavailable slots to keep from deleting more new pods.

If we're looking to backport this, that might be a minimal place to start (and test the scenario I described), then maybe follow up with some variable renames in master only to make the bookkeeping clearer.

Reply:
after both @soltysh and I failed to produce a unit test that would reproduce this case, I read closer, and the `numUnavailable >= maxUnavailable` case does a `continue`, not a `break`, so it just continues to the next pod and unavailable old pods keep accumulating in `allowedReplacementPods`... sorry for misreading the loop logic 🤦